File: topInterview150/array-string/bestTimeToBuyAndSellStock.go
package main

import "math"

func maxProfit(prices []int) int {
	maxIndex, minIndex := 0, 0
	max := 0
	min := math.MaxInt
	for i := 0; i < len(prices); i++ {
		if prices[i] < min {
			min = prices[i]
			minIndex = i
		}
	}
	for i := minIndex; i < len(prices); i++ {
		if prices[i] > max {
			max = prices[i]
			maxIndex = i
		}
	}
	if maxIndex == 0 {
		return 0
	}
	return max - min
}

File: neetcode-roadmap/sliding-window/BestTimeToBuyAndSellStock.java
class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int left = 0;
        int right = 1;

        while (right < prices.length){
            int profit;
            if (prices[left] < prices[right]){
                profit = prices[right] - prices[left];
                maxProfit = Math.max(maxProfit, profit);
            }else {
                left = right;
            }
            right++;
        }
        return maxProfit;
    }
}